app/scripts/controllers/network/createInfuraClient.test.js

```diff
@@ -0,0 +1,830 @@
+  { method, params = [] },
+  blockParamIndex,
+  blockParam,
+) {
+  const updatedParams = params.slice();
+  updatedParams[blockParamIndex] = blockParam;
+  return { method, params: updatedParams };
+}
+
+describe('createInfuraClient', () => {
+  // The first time an RPC method is requested, the latest block number is
+  // pulled from the block tracker, the RPC method is delegated to Infura, and
+  // the result is cached under that block number, as long as the result is
+  // "non-empty". The next time the same request takes place, Infura is not hit,
+  // and the result is pulled from the cache.
+  //
+  // For most RPC methods here, a "non-empty" result is a result that is not
+  // null, undefined, or a non-standard "nil" value that geth produces.
+  //
+  // Some RPC methods are cacheable. Consult the definitive list of cacheable
+  // RPC methods in `cacheTypeForPayload` within `eth-json-rpc-middleware`.
+
+  describe('when the RPC method is eth_chainId', () => {
+    it('does not hit Infura, instead returning the chain id that maps to the Infura network, as a hex string', async () => {
+      const chainId = await withInfuraClient(
+        { network: 'ropsten' },
+        ({ makeRpcCall }) => {
+          return makeRpcCall({
+            method: 'eth_chainId',
+          });
+        },
+      );
+
+      expect(chainId).toStrictEqual('0x3');
+    });
+  });
+
+  describe('when the RPC method is net_version', () => {
+    it('does not hit Infura, instead returning the chain id that maps to the Infura network, as a decimal string', async () => {
+      const chainId = await withInfuraClient(
+        { network: 'ropsten' },
+        ({ makeRpcCall }) => {
+          return makeRpcCall({
+            method: 'net_version',
+          });
+        },
+      );
+
+      expect(chainId).toStrictEqual('3');
+    });
+  });
+
+  // == RPC methods that do not support params
+  //
+  // For `eth_getTransactionByHash` and `eth_getTransactionReceipt`, a
+  // "non-empty" result is not only one that is not null, undefined, or a
+  // geth-specific "nil", but additionally a result that has a `blockHash` that
+  // is not 0x0.
+
+  describe.each([
+    'eth_blockNumber',
+    'eth_compileLLL',
+    'eth_compileSerpent',
+    'eth_compileSolidity',
+    'eth_estimateGas',
+    'eth_gasPrice',
+    'eth_getBlockByHash',
+    'eth_getBlockTransactionCountByHash',
+    'eth_getBlockTransactionCountByNumber',
+    'eth_getCompilers',
+    'eth_getFilterLogs',
+    'eth_getLogs',
+    'eth_getTransactionByBlockHashAndIndex',
+    'eth_getTransactionByBlockNumberAndIndex',
+    'eth_getUncleByBlockHashAndIndex',
+    'eth_getUncleByBlockNumberAndIndex',
+    'eth_getUncleCountByBlockHash',
+    'eth_getUncleCountByBlockNumber',
+    'eth_protocolVersion',
+    'shh_version',
+    'test_blockCache',
+    'test_forkCache',
+    'test_permaCache',
+    'web3_clientVersion',
+    'web3_sha3',
+  ])('when the RPC method is %s', (method) => {
```

Comment:
TIL that `describe.each` exists

Reply:
I decided not to use this as it made the tests less easy to extend later.